experimental/formatted_string/src/formatted_string.rs

```diff
@@ -0,0 +1,219 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::FormattedStringError;
+use alloc::borrow::ToOwned;
+use alloc::vec::Vec;
+use core::str;
+
+#[derive(Copy, Clone, PartialEq, Debug)]
+pub enum LocationInPart {
+    Begin,
+    Extend,
+}
+
+pub trait FormattedStringLike<F: Copy, const L: usize>: AsRef<str> {
+    fn fields_at(&self, pos: usize) -> [F; L] {
+        self.annotation_at(pos).map(|(_, field)| field)
+    }
+
+    fn is_field_start(&self, pos: usize, level: usize) -> bool {
+        assert!(level < L);
+        let (location, _) = self.annotation_at(pos)[level];
+        location == LocationInPart::Begin
+    }
+
+    #[doc(hidden)]
+    // This is used to make the inserts more efficient; clients should
+    // use fields_at and is_field_start.
+    fn annotation_at(&self, pos: usize) -> &[(LocationInPart, F); L];
+}
+
+impl<F: Copy> FormattedStringLike<F, 0> for &str {
+    fn annotation_at(&self, _pos: usize) -> &[(LocationInPart, F); 0] {
+        // Yay we can return dangling references for singleton types!
+        &[]
+    }
+}
+
+/// A string with L levels of formatting annotations.
+#[derive(Debug, PartialEq)]
+pub struct FormattedString<F: Copy, const L: usize> {
+    // bytes is always valid UTF-8, so from_utf8_unchecked is safe
+    bytes: Vec<u8>,
+    // The first dimension corresponds to the bytes, the second are the L levels of annotations
+    annotations: Vec<[(LocationInPart, F); L]>,
+}
+
+impl<F: Copy, const L: usize> AsRef<str> for FormattedString<F, L> {
+    fn as_ref(&self) -> &str {
+        unsafe { &str::from_utf8_unchecked(&self.bytes) }
+    }
+}
+
+impl<F: Copy, const L: usize> FormattedStringLike<F, L> for FormattedString<F, L> {
+    fn annotation_at(&self, pos: usize) -> &[(LocationInPart, F); L] {
+        &self.annotations[pos]
+    }
+}
+
+impl<F: Copy, const L: usize> FormattedString<F, L> {
```

Comment:
Note: I see you went back on what you had originally proposed, which was something like `MultiLevelFormattedString<N>` and `FormattedString` which is an alias for the larger type with N=1.  Why the change of heart?  (I think it's okay either way)